node/service/src/lib.rs

```diff
@@ -1208,6 +1268,77 @@ where
 	})
 }
 
+/// Get the workers path by considering the passed-in `workers_path` option, or possible locations
+/// on the filesystem. See `new_full`.
+fn get_workers_paths(
+	given_workers_path: Option<std::path::PathBuf>,
+) -> Result<Vec<(std::path::PathBuf, std::path::PathBuf)>, Error> {
+	if let Some(path) = given_workers_path {
+		log::trace!("Using explicitly provided workers path {:?}", path);
+
+		if path.is_executable() {
+			return Ok(vec![(path.clone(), path)])
+		}
+
+		let mut prep_worker = path.clone();
+		let mut exec_worker = path.clone();
+		prep_worker.push(polkadot_node_core_pvf::PREPARE_BINARY_NAME);
+		exec_worker.push(polkadot_node_core_pvf::EXECUTE_BINARY_NAME);
+
+		// Check if both workers exist. Otherwise return an empty vector which results in an error.
+		return if prep_worker.exists() && exec_worker.exists() {
+			Ok(vec![(prep_worker, exec_worker)])
+		} else {
+			Ok(vec![])
+		}
+	}
+
+	// Workers path not provided, check all possible valid locations.
+
+	let mut workers_paths = vec![];
+
+	{
+		let mut exe_path = std::env::current_exe()?;
+		let _ = exe_path.pop(); // executable file will always have a parent directory.
+		let mut prep_worker = exe_path.clone();
+		prep_worker.push(polkadot_node_core_pvf::PREPARE_BINARY_NAME);
+		let mut exec_worker = exe_path.clone();
+		exec_worker.push(polkadot_node_core_pvf::EXECUTE_BINARY_NAME);
+
+		// Add to set if both workers exist. Warn on partial installs.
+		let (prep_worker_exists, exec_worker_exists) = (prep_worker.exists(), exec_worker.exists());
+		if prep_worker_exists && exec_worker_exists {
+			log::trace!("Worker binaries found at current exe path: {:?}", exe_path);
+			workers_paths.push((prep_worker, exec_worker));
+		} else if prep_worker_exists {
+			log::warn!("Worker binary found at {:?} but not {:?}", prep_worker, exec_worker);
+		} else if exec_worker_exists {
+			log::warn!("Worker binary found at {:?} but not {:?}", exec_worker, prep_worker);
+		}
+	}
+
+	{
+		let libexec = PathBuf::from("/usr/libexec");
```

Comment:
I think I'd prefer using e.g. `/usr/lib/polkadot` instead here (which is also allowed according to LSB); `/usr/libexec` is actually not that commonly used in practice. (e.g. on my workstation I don't even have this directory, and on my personal server that I maintain it exists but only contains two files, while `/usr/lib` contains many more.)

Reply:
Sounds good to me. Some background from the [manual](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s07.html):

> Some previous versions of this document did not support /usr/libexec, despite it being standard practice in a number of environments. [26] To accomodate this restriction, it became common practice to use /usr/lib instead. Either practice is now acceptable, but each application must choose one way or the other to organize itself.